crypto/fipsmodule/service_indicator/service_indicator.c

```diff
@@ -217,6 +217,44 @@ static int is_md_fips_approved_for_verifying(int md_type, int pkey_type) {
   }
 }
 
+// custom_meth_invoked checks whether custom crypto was invoked in the |meth|
+// or |eckey_method| fields for a given |RSA| or |EC_KEY| respectively. For
+// |RSA| keys, custom verify and sign functionality is supported. For |EC_KEY|
+// keys, only custom sign functionality is supported.
+// Returns one if custom crypto was invoked and zero otherwise.
+static int custom_meth_invoked(const EVP_PKEY_CTX *ctx) {
+  const int pkey_type = EVP_PKEY_id(ctx->pkey);
+  switch (pkey_type) {
+    case EVP_PKEY_RSA:
+    case EVP_PKEY_RSA_PSS: {
+      const RSA_METHOD *meth = ctx->pkey->pkey.rsa->meth;
+      // Must be either |EVP_PKEY_OP_VERIFY| or |EVP_PKEY_OP_SIGN|
+      switch (ctx->operation) {
+        case EVP_PKEY_OP_VERIFY:
+          return meth->verify_raw ? 1 : 0;
+
+        case EVP_PKEY_OP_SIGN:
+          // There are cases where custom |sign| functionality may be set but
+          // not |sign_raw|. This check is more conservative and fails if
+          // custom functionality is provided for either function pointer.
+          return (meth->sign || meth->sign_raw) ? 1 : 0;
+
+        default:
+          return 0;  // custom crypto can't be invoked for unsupported ops
+      }
+    }
+
+    case EVP_PKEY_EC: {
+      const EC_KEY_METHOD *meth = ctx->pkey->pkey.ec->eckey_method;
+      return (meth->sign || meth->sign_sig) ? 1 : 0;
+    }
```

Comment:
This should only (potentially) return 1 for an `EVP_PKEY_EC` if the operation is `EVP_PKEY_OP_SIGN`:
```
    case EVP_PKEY_EC: {
      const EC_KEY_METHOD *meth = ctx->pkey->pkey.ec->eckey_method;
      // Only custom sign functionality is supported.
      if (ctx->operation == EVP_PKEY_OP_SIGN) {
        return (meth->sign || meth->sign_sig) ? 1 : 0;
      }
      return 0;
    }
```